### PR TITLE
geometry utilities: weighted_centroid_points; Line.point; Polyline.point

### PR DESCRIPTION
**Pull request by robin-oval**

### CODE REVIEW

File: src/compas/geometry/average.py [hunk 1]
```diff
@@ -58,6 +60,25 @@ def centroid_points(points):
...
+def weighted_centroid_points(points, weights):
+    """Compute the weighted centroid of a set of points.
+
+    Parameters
+    ----------
+    points : list
+        A list of point coordinates.
+    weights : list
+        A list of weight floats.
+
```
Comment (tomvanmele) [inline src/compas/geometry/average.py:160]: are there no bounds on the weights?

File: src/compas/geometry/objects/line.py [hunk 2]
```diff
@@ -144,6 +144,17 @@ def __iter__(self):
...
 
+    # ==========================================================================
+    # queries
+    # ==========================================================================
+
+    def point(self, t):
+        """Point: The point from the start to the end at a specific normalized parameter."""
+        if t < 0 or t > 1:
+            return None
+        v = self.direction * (t * self.length)
```
Comment (tomvanmele) [inline src/compas/geometry/objects/line.py:163]: should we not also check for the bounds and then return one of the points directly if appropriate?

File: src/compas/geometry/objects/polyline.py [hunk 3]
```diff
@@ -128,6 +130,43 @@ def __iter__(self):
...
+    # queries
+    # ==========================================================================
+
+    def point(self, t, snap = False):
+        """Point: The point from the start to the end at a specific normalized parameter.
+        If snap is True, return the closest polyline point."""
+
+        if t < 0 or t > 1:
+            return None
+
```
Comment (tomvanmele) [inline src/compas/geometry/objects/polyline.py]: check for boundary values and return points immediately if appropriate?

File: src/compas/geometry/average.py (hunk 1 above)
Comment (robin-oval) [reply src/compas/geometry/average.py:160]: No they weights can be negative. I added a comment to mention it.

File: src/compas/geometry/objects/line.py (hunk 2 above)
Comment (robin-oval) [reply src/compas/geometry/objects/line.py:163]: I imagine one would use line.start or .end directly for the bounds? The method is still valid for the extremities. But I can add something like if you wish.
`if t == 0:
return self.start
elif t==1:
return.end`

File: src/compas/geometry/objects/polyline.py (hunk 3 above)
Comment (robin-oval) [reply src/compas/geometry/objects/polyline.py]: Same thing as for the Line class.
I could also create new methods like:
'def start(self):
return self.points[0]
def end(self):
return self.points[-1]'

**Comment by robin-oval** [comment]
should I add the little additions I mention or is it not necessary?

**Comment by tomvanmele** [comment]
what do you mean

File: src/compas/geometry/average.py (hunk 1 above)
Comment (tomvanmele) [reply src/compas/geometry/average.py:160]: but the weights are not bound to an interval between 0 and +/- 1?

**Comment by robin-oval** [comment]
returning directly the bounds if t=0 or 1, though not necessary to add

File: src/compas/geometry/objects/line.py (hunk 2 above)
Comment (tomvanmele) [reply src/compas/geometry/objects/line.py:163]: would make sense in my opinion

File: src/compas/geometry/average.py (hunk 1 above)
Comment (robin-oval) [reply src/compas/geometry/average.py:160]: no the weights have no bounds

File: src/compas/geometry/objects/line.py (hunk 2 above)
Comment (robin-oval) [reply src/compas/geometry/objects/line.py:163]: ok

**Comment by tomvanmele** [comment]
after the small updates this is ready to be merged...

**Comment by robin-oval** [comment]
done!

**Comment by tomvanmele** [comment]
we should also start writing tests for all of these things.
will add a template soon...

**Pull request merged**